doc/source-highlight.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH SOURCE-HIGHLIGHT "1" "July 2009" "Source-highlight 3.1.1 (library: 3:0:2)" "User Commands"
.SH NAME
Source-highlight \- convert source code to syntax highlighted document
.SH SYNOPSIS
.B source-highlight
[\fIOPTIONS\fR]... \fI< input_file > output_file\fR
.SH DESCRIPTION
GNU source\-highlight 3.1.1
.PP
Highlight the syntax of a source file (e.g. Java) into a specific format (e.g.
HTML)
.IP
source\-highlight [OPTIONS]... \fB\-i\fR input_file \fB\-o\fR output_file
source\-highlight [OPTIONS]... [FILES]...
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help and exit
.TP
\fB\-\-detailed\-help\fR
Print help, including all details and hidden
options, and exit
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version and exit
.TP
\fB\-i\fR, \fB\-\-input\fR=\fIfilename\fR
input file. default std input
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIfilename\fR
output file. default std output. If STDOUT is
specified, the output is directed to standard
output
.TP
\fB\-s\fR, \fB\-\-src\-lang\fR=\fISTRING\fR
source language (use \fB\-\-lang\-list\fR to get the
complete list).  If not specified, the source
language will be guessed from the file
extension.
.TP
\fB\-\-lang\-list\fR
list all the supported language and associated
language definition file
.TP
\fB\-\-outlang\-list\fR
list all the supported output language and
associated language definition file
.TP
\fB\-f\fR, \fB\-\-out\-format\fR=\fISTRING\fR
output format (use \fB\-\-outlang\-list\fR to get the
complete list)  (default=`html')
.TP
\fB\-d\fR, \fB\-\-doc\fR
create an output file that can be used as a
stand alone document (e.g., not to be
included in another one)
.TP
\fB\-\-no\-doc\fR
cancel the \fB\-\-doc\fR option even if it is implied
(e.g., when css is given)
.TP
\fB\-c\fR, \fB\-\-css\fR=\fIfilename\fR
the external style sheet filename.  Implies
\fB\-\-doc\fR
.TP
\fB\-T\fR, \fB\-\-title\fR=\fISTRING\fR
give a title to the output document.  Implies
\fB\-\-doc\fR
.TP
\fB\-t\fR, \fB\-\-tab\fR=\fIINT\fR
specify tab length.  (default=`8')
.TP
\fB\-H\fR, \fB\-\-header\fR=\fIfilename\fR
file to insert as header
.TP
\fB\-F\fR, \fB\-\-footer\fR=\fIfilename\fR
file to insert as footer
.TP
\fB\-\-style\-file\fR=\fIfilename\fR
specify the file containing format options
(default=`default.style')
.TP
\fB\-\-style\-css\-file\fR=\fIfilename\fR specify the file containing format options (in
css syntax)
.TP
\fB\-\-style\-defaults\fR=\fIfilename\fR specify the file containing defaults for format
options  (default=`style.defaults')
.TP
\fB\-\-outlang\-def\fR=\fIfilename\fR
output language definition file
.TP
\fB\-\-outlang\-map\fR=\fIfilename\fR
output language map file
(default=`outlang.map')
.TP
\fB\-\-data\-dir\fR=\fIpath\fR
directory where language definition files and
language maps are searched for.  If not
specified these files are searched for in the
current directory and in the data dir
installation directory
.TP
\fB\-\-output\-dir\fR=\fIpath\fR
output directory
.TP
\fB\-\-lang\-def\fR=\fIfilename\fR
language definition file
.TP
\fB\-\-lang\-map\fR=\fIfilename\fR
language map file  (default=`lang.map')
.TP
\fB\-\-show\-lang\-elements\fR=\fIfilename\fR
prints the language elements that are defined
.IP
in the language definition file
.TP
\fB\-\-infer\-lang\fR
force to infer source script language
(overriding given language specification)
.SS "Lines:"
.TP
\fB\-n\fR, \fB\-\-line\-number\fR[=\fIpadding\fR]
number all output lines, using the specified
padding character  (default=`0')
.TP
\fB\-\-line\-number\-ref\fR[=\fIprefix\fR]
number all output lines and generate an anchor,
.IP
made of the specified prefix + the line
number  (default=`line')
.PP
Filtering output:
.IP
Mode: linerange
.IP
specifying line ranges
.TP
\fB\-\-line\-range\fR=\fISTRING\fR
generate only the lines in the specified
range(s)
.TP
\fB\-\-range\-separator\fR=\fISTRING\fR
the optional separator to be printed among
ranges (e.g., "...")
.TP
\fB\-\-range\-context\fR=\fIINT\fR
number of (context) lines generated even if not
in range
.IP
Mode: regexrange
.IP
specifying regular expression delimited ranges
.TP
\fB\-\-regex\-range\fR=\fISTRING\fR
generate only the lines within the specified
regular expressions
.SS "reference generation:"
.TP
\fB\-\-gen\-references\fR=\fISTRING\fR
generate references  (possible
values="inline", "postline", "postdoc"
default=`inline')
.TP
\fB\-\-ctags\-file\fR=\fIfilename\fR
specify the file generated by ctags that will
be used to generate references
(default=`tags')
.TP
\fB\-\-ctags\fR=\fIcmd\fR
how to run the ctags command.  If this option
is not specified, ctags will be executed with
the default value.  If it is specified with
an empty string, ctags will not be executed
at all  (default=`ctags \fB\-\-excmd\fR=\fIn\fR
\fB\-\-tag\-relative\fR=\fIyes\fR')
.SS "testing:"
.TP
\fB\-v\fR, \fB\-\-verbose\fR
verbose mode on
.TP
\fB\-q\fR, \fB\-\-quiet\fR
print no progress information
.TP
\fB\-\-binary\-output\fR
write output files in binary mode
.TP
\fB\-\-statistics\fR
print some statistics (i.e., elapsed time)
.TP
\fB\-\-gen\-version\fR
put source\-highlight version in the generated
file  (default=on)
.TP
\fB\-\-check\-lang\fR=\fIfilename\fR
only check the correctness of a language
definition file
.TP
\fB\-\-check\-outlang\fR=\fIfilename\fR
only check the correctness of an output
language definition file
.TP
\fB\-\-failsafe\fR
if no language definition is found for the
input, it is simply copied to the output
.TP
\fB\-g\fR, \fB\-\-debug\-langdef\fR[=\fItype\fR]
debug a language definition.  In dump mode just
dumps all the steps; in interactive, at each
step, waits for some input (press ENTER to
step)  (possible values="interactive",
"dump" default=`dump')
.TP
\fB\-\-show\-regex\fR=\fIfilename\fR
show the regular expression automaton
corresponding to a language definition file
.PP
Maintained by Lorenzo Bettini <http://www.lorenzobettini.it>
.SH "REPORTING BUGS"
Report bugs to <bug\-source\-highlight at gnu.org>
.SH COPYRIGHT
Copyright \(co 1999-2008 Lorenzo Bettini <http://www.lorenzobettini.it>
This program comes with ABSOLUTELY NO WARRANTY.
.br
This is free software; you may redistribute copies of the program
under the terms of the GNU General Public License.
For more information about these matters, see the file named COPYING.
.SH "SEE ALSO"
The full documentation for
.B Source-highlight
is maintained as a Texinfo manual.  If the
.B info
and
.B Source-highlight
programs are properly installed at your site, the command
.IP
.B info Source-highlight
.PP
should give you access to the complete manual.


NEWS
source-highlight -- History of visible changes.

Copyright (C) 1999-2009 Lorenzo Bettini
http://www.lorenzobettini.it
http://www.gnu.org/software/src-highlite

See the end for copying conditions.

Please send source-highlight bug reports to bug-source-highlight at gnu.org

Version 3.1.1

   * access to static global LangDefManager, LangMaps is provided
     through the class Instances
   * mechanism for setting a global data dir value (e.g., to ensure
     consistency in a library) and for checking settings
   * library utility functions to retrieve .lang and .outlang files
   * fixed default output format to "html.outlang" not "html.lang"
   * SourceHighlight class return the output file extension
   * the background color for the document is kept empty if it is
     not specified in the style file
   * SourceHighlight.setLineNumberDigits: to
	 explicitly set the number of digits for line numbers
   * utils.h file is installed in the header directory
   * white is a standard color in style files
   * language definition file for Texinfo
   * language definition file for Haskell (thanks to Sergey Astanin)

Version 3.1

   * library utility functions to retrieve .style and .css files
   * fixed library manual link in index.html
   * doxyfile is now distributed
   * better formatting for < and > in latex output
   * doxygen documentation can be built even when building in
     a separate directory
   * added man page for source-highlight-settings
   * language definition for Oz, thanks to David Bremners

Version 3.0

   * provides a library for highlighting that can be used
     by other programs.
   *  --regex-range (highlight only specified lines of an input file,
     according to specific regular expressions describing the delimiting
     lines).
   * deal with HTML5 features (thanks to Simon Pieters)
   * use --docdir for documentation directory
   * --binary-output command line option to write output files in binary
     mode
   * Added the program source-highlight-settings to write a configuration
     file for source-highlight (and its library) in the user home directory.
   * languange definition for generic configuration files
   * languange definition for pkg-config files
   * the input language is now discovered also using the whole file
     name (besides the file extension if any)
   * in .lang files one can specify the exit level

Version 2.11.1

   * fixed an installation bug in non-root mode (thanks to Britton Kerin)
   * fortran.lang improved with new keywords from Fortran 90 
     through Fortran 2003 (thanks to Jason Blevins)
   * javascript.lang improved (thanks to gnombat)
   * added keywords to caml language definition
   * Ada language definition file (thanks to Reto Buerki)
   * lang and outlang files can also be downloaded independently from
     the source-highlight site

Version 2.11

   * optimized the procedure for finding the best matching expression
   * improved boost regex library search in configure script
   * fixed regexp element in javascript lang definition
   * highlight types used in variable/parameter declarations in C++ and Java
   * fixed language def files for properties and desktop files
   * newline chars can be inserted (and ignored) in regular expression
     specifications
   * --line-range (highlight only specified lines of an input file,
     --range-context to specify the number of surrounding context lines,
     --range-separator the separator line between ranges)
   * language definition for Xorg configuration files
   * language definition for Scala (thanks to Dean Wampler)
   * bash completion file for source-highlight (thanks to Tiziano Muller)

Version 2.10

   * formatting is applied even when generating anchors and references
   * noref is handled also for output languages using onestyle (e.g., xhtml)
   * improved fortran highlighting
   * improved python highlighting
   * fixed title in docbook output
   * language definition for ldap files (e.g., ldiff files)
   * language definition for autoconf files
   * improved m4 language definition
   * improved logtalk language definition
   * url.lang handles ~ in urls
   * language definition for glsl.lang (provided by Cesare Tirabassi)

Version 2.9

   * language definition for C (not C++) files
   * language definition for properties files
   * language definition for KDE desktop and ini files
   * language definition for lsm files (Linux Software Map)
   * language definition for rpm spec files
   * language definition for Haxe files (thanks to Jos Hirth)
   * style.defaults for associating a style for an element (whose style is
     not specified) to the style of another element
   * highlight some KDE programming files (e.g., .rc, .kcfg, etc.)
   * correctly highlight for less when filenames contain paths
   * fixed a bug in file inclusion of langdef files
   * fixed compilation problems for fileutil.cc (thanks to Adrian Reber)
   * xml elements are correctly recognized when containing .
     (thanks to Toby White)
   * references for xhtml output files

Version 2.8

   * lang definition for slang (by John E. Davis)
   * correctly handle words in ' and ` regular expression strings
   * the right delimiter of a delim element can refer to marked subexpressions
     in the left delimiter
   * fixed the definition of C-style comments which are not nested
   * improved perl syntax highlighting
   * javascript regular expression highlighting
   * padding character for line numbers can be specified (thanks to
     Roger Nilsson)
   * removed non standard % make rules

Version 2.7


src/source-highlight.cc
/*
 * Copyright (C) 1999-2009  Lorenzo Bettini <http://www.lorenzobettini.it>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <cstdlib>
#include <iostream>
#include <exception>
#include <boost/shared_ptr.hpp>

#include "srchilite/sourcehighlight.h"

#include "srchilite/copyright.h"
#include "srchilite/reportbugs.h"
#include "srchilite/fileutil.h"
#include "srchilite/verbosity.h"
#include "srchilite/langmap.h"
#include "srchilite/languageinfer.h"
#include "srchilite/utils.h"
#include "srchilite/highlightbuilderexception.h"
#include "srchilite/parserexception.h"
#include "srchilite/debuglistener.h"
#include "srchilite/ctagsmanager.h"
#include "srchilite/stopwatch.h"
#include "srchilite/lineranges.h"
#include "srchilite/regexranges.h"
#include "srchilite/versions.h"
#include "srchilite/settings.h"

#include "cmdline.h"

#include "progname.h"

using namespace std;
using namespace srchilite;

/**
 * how language inference should be made: NOINFERENCE = no inference at all,
 * INFERFIRST = inference has priority (e.g., if --infer-lang is specified),
 * INFERATLAST = try inference as the last chance
 */
enum InferPolicy {
    NOINFERENCE = 0, INFERFIRST, INFERATLAST
};

static void print_copyright();
static void print_reportbugs();
static void print_version();
static string inferLang(const string &inputFileName);
static string getMappedLang(const string &s, LangMap &langmap);
static void printError(const string &s);
static void print_cgi_header();

/**
 * Exits for a generic error (i.e., file not found)
 */
static void exitError(const string &s);

static string getLangFileName(InferPolicy infer, const string &inputFileName,
        const string &langFileName, const string &langName, LangMap &langMap);

static bool verbose = false;
static bool failsafe = false;

#ifdef BUILD_AS_CGI
#include "envmapper.h"
#endif // BUILD_AS_CGI
/**
 * Print progress status information (provided --quiet is not specified)
 * @param message
 */
#define PROGRESSINFO(message) if (!args_info.quiet_given) cerr << message;

int main(int argc, char * argv[]) {
    set_program_name(argv[0]);
    gengetopt_args_info args_info; // command line structure
    bool is_cgi = false;

#ifdef BUILD_AS_CGI
    // map environment to parameters if used as CGI
    char **temp_argv;
    temp_argv = map_environment(&argc, argv);
    is_cgi = temp_argv != argv;
    argv = temp_argv;
#endif // BUILD_AS_CGI
    if (cmdline_parser(argc, argv, &args_info) != 0) {
        // calls cmdline parser. The user gived bag args if it doesn't return -1
        return EXIT_FAILURE;
    }

    if (args_info.version_given) {
        print_version();
        print_copyright();
        cmdline_parser_free(&args_info);
        return EXIT_SUCCESS;
    }

    if (args_info.help_given || args_info.detailed_help_given) {
        cout << "GNU ";
        if (args_info.detailed_help_given) {
            cmdline_parser_print_detailed_help();
        } else {
            cmdline_parser_print_help();
        }
        print_reportbugs();
        cmdline_parser_free(&args_info);
        return EXIT_SUCCESS;
    }

    // set possible line ranges
    boost::shared_ptr<LineRanges> lineRanges;
    if (args_info.line_range_given) {
        lineRanges = boost::shared_ptr<LineRanges>(new LineRanges);
        for (unsigned int i = 0; i < args_info.line_range_given; ++i) {
            if (lineRanges->addRange(args_info.line_range_arg[i]) != NO_ERROR) {
                string invalid_range = args_info.line_range_arg[i];
                cmdline_parser_free(&args_info);
                exitError("invalid line range: " + invalid_range);
            }
        }
    }

    // set possible regex ranges
    boost::shared_ptr<RegexRanges> regexRanges;
    if (args_info.regex_range_given) {
        regexRanges = boost::shared_ptr<RegexRanges>(new RegexRanges);
        for (unsigned int i = 0; i < args_info.regex_range_given; ++i) {
            if (!regexRanges->addRegexRange(args_info.regex_range_arg[i])) {
                string invalid_range = args_info.regex_range_arg[i];
                cmdline_parser_free(&args_info);
                exitError("invalid regex range: " + invalid_range);
            }
        }
    }

    if (args_info.range_context_given) {
        lineRanges->setContextLines(args_info.range_context_arg);
    }

    // if a stopwatch is created, when it is deleted (automatically
    // since we're using a shared pointer, it will print the
    // elapsed seconds.
    boost::shared_ptr<StopWatch> stopwatch;
    if (args_info.statistics_given) {
        stopwatch = boost::shared_ptr<StopWatch>(new StopWatch);
    }

    verbose = args_info.verbose_given;
    failsafe = args_info.failsafe_given;

    string inputFile;
    string outputFile;
    string srcLang;
    string outputFormat;
    string langFile;
    string outLangFile;
    string dataDir;
    string outputDir;

    Verbosity::setVerbosity(verbose);

    if (args_info.input_given)
        inputFile = args_info.input_arg;

    if (args_info.output_given)
        outputFile = args_info.output_arg;

    // for cgi force output to standard output
    if (is_cgi)
        outputFile = "STDOUT";

    if (args_info.src_lang_given)
        srcLang = args_info.src_lang_arg;

    if (args_info.lang_def_given)
        langFile = args_info.lang_def_arg;

    if (args_info.outlang_def_given)
        outLangFile = args_info.outlang_def_arg;

    // the default for output format is html
    outputFormat = args_info.out_format_arg;

    if (args_info.data_dir_given) {
        dataDir = args_info.data_dir_arg;
    }

    if (args_info.output_dir_given) {
        outputDir = args_info.output_dir_arg;
    }

    /*
     the starting default path to search for files is computed at
     run-time: it is
     the path of the binary + ".." + RELATIVEDATADIR
     this should make the package relocable (i.e., not stuck
     with a fixed installation directory).
     Of course, the GNU standards for installation directories
     should be followed, but this is not a problem if you use
     configure and make install features.
     If no path is specified in the running program we go back to
     the absolute datadir.
     */
    // this is defined in fileutil.cc
    string prefix_dir = get_file_path(argv[0]);
    if (prefix_dir.size())
        start_path = get_file_path(argv[0]) + RELATIVEDATADIR;
    else
        start_path = Settings::retrieveDataDir();

    // if datadir is not specified, we rely on start_path?

    try {
        // initialize map files
        LangMap langmap(dataDir, args_info.lang_map_arg);

        // just print the input language list and associations
        if (args_info.lang_list_given) {
            langmap.open();
            langmap.print();
            return (EXIT_SUCCESS);
        }

        LangMap outlangmap(dataDir, args_info.outlang_map_arg);

        // just print the input language list and associations
        if (args_info.outlang_list_given) {
            outlangmap.open();
            outlangmap.print();
            return (EXIT_SUCCESS);
        }

        string cssUrl;
        if (args_info.css_given)
            cssUrl = args_info.css_arg;

        string docTitle;
        if (args_info.title_given)
            docTitle = args_info.title_arg;

        if (args_info.check_lang_given || args_info.check_outlang_given
                || args_info.show_regex_given
                || args_info.show_lang_elements_given
                || args_info.lang_list_given || args_info.outlang_list_given) {
            // FIXME just a bogus outlang file specification not to search in outlang.map
            outLangFile = "html.outlang";
        }

        // initialize the main source highlight object
        SourceHighlight sourcehighlight(getLangFileName(NOINFERENCE,
                outputFile, outLangFile, outputFormat
                        + (cssUrl.size() ? +"-css" : ""), outlangmap));

        // and sets all its properties according to the command line args
        sourcehighlight.setDataDir(dataDir);

        boost::shared_ptr<DebugListener> debugListener;

        // if a simple check is required...
        if (args_info.check_lang_given) {
            cout << "checking " << args_info.check_lang_arg << "... ";
            sourcehighlight.checkLangDef(args_info.check_lang_arg);
            // if we're here, everything went fine!
            cout << "OK" << endl;
            // otherwise an exception has already been thrown
            return (EXIT_SUCCESS);
        }

        if (args_info.check_outlang_given) {
            cout << "checking " << args_info.check_outlang_arg << "... ";
            sourcehighlight.checkOutLangDef(args_info.check_outlang_arg);
            // if we're here, everything went fine!
            cout << "OK" << endl;
            // otherwise an exception has already been thrown
            return (EXIT_SUCCESS);
        }

        // just print the highlightstate if show-regex is specified
        if (args_info.show_regex_given) {
            sourcehighlight.printHighlightState(args_info.show_regex_arg, cout);
            return (EXIT_SUCCESS);
        }

        // just print the lang elems if show-lang-elements is specified
        if (args_info.show_lang_elements_given) {
            sourcehighlight.printLangElems(args_info.show_lang_elements_arg,
                    cout);
            return (EXIT_SUCCESS);
        }

        sourcehighlight.setGenerateEntireDoc((!args_info.no_doc_given)
                && (args_info.doc_given || (docTitle.size()) || cssUrl.size()));
        sourcehighlight.setStyleDefaultFile(args_info.style_defaults_arg);
        sourcehighlight.setStyleFile(args_info.style_file_arg);
        if (args_info.style_css_file_given)
            sourcehighlight.setStyleCssFile(args_info.style_css_file_arg);
        sourcehighlight.setGenerateVersion(args_info.gen_version_flag != 0);
        sourcehighlight.setCss(cssUrl);
        sourcehighlight.setTitle(docTitle);
        sourcehighlight.setOutputDir(outputDir);
        sourcehighlight.setLineRanges(lineRanges.get());
        sourcehighlight.setRegexRanges(regexRanges.get());
        sourcehighlight.setBinaryOutput(args_info.binary_output_given);

        if (args_info.debug_langdef_given) {
            debugListener = boost::shared_ptr<DebugListener>(new DebugListener);
            const string debugType = args_info.debug_langdef_arg;
            if (debugType == "interactive")
                debugListener->setInteractive(true);
            sourcehighlight.setHighlightEventListener(debugListener.get());
            // during debugging disable optimizations
            sourcehighlight.setOptimize(false);
        }

        if (args_info.header_given)
            sourcehighlight.setHeaderFileName(args_info.header_arg);

        if (args_info.footer_given)
            sourcehighlight.setFooterFileName(args_info.footer_arg);

        bool generateLineNumbers = args_info.line_number_given
                || args_info.line_number_ref_given;
        sourcehighlight.setGenerateLineNumbers(generateLineNumbers);

        if (args_info.line_number_given) {
            // set the line number padding char (default is always '0')
            sourcehighlight.setLineNumberPad(args_info.line_number_arg[0]);
        }

        if (args_info.line_number_ref_given) {
            sourcehighlight.setGenerateLineNumberRefs(true);
            sourcehighlight.setLineNumberAnchorPrefix(
                    args_info.line_number_ref_arg);
        }

        if (generateLineNumbers) {
            // when generating line numbers tabs must be translated
            // otherwise the output line will not be aligned
            // due to the presence of line numbers
            sourcehighlight.setTabSpaces(8);
        }

        // whether to give precedence to language inference
        InferPolicy inferPolicy = (args_info.infer_lang_given ? INFERFIRST
                : INFERATLAST);

        RefPosition refposition;
        string gen_references_arg = args_info.gen_references_arg;
        if (gen_references_arg == "inline")
            refposition = INLINE;
        else if (gen_references_arg == "postline")
            refposition = POSTLINE;
        else if (gen_references_arg == "postdoc")
            refposition = POSTDOC;

        boost::shared_ptr<CTagsManager> ctagsManager;
        if (args_info.gen_references_given) {
            string ctags = args_info.ctags_arg;
            string ctags_file = args_info.ctags_file_arg;

            // build the additional arguments for the ctags command
            if (args_info.gen_references_given && ctags != "") {
                if (inputFile.size()) {
                    ctags += " ";
                    ctags += inputFile;
                } else if (args_info.inputs_num) {
                    for (unsigned int i = 0; i < (args_info.inputs_num); ++i) {
                        ctags += " ";
                        ctags += args_info.inputs[i];
                    }
                }
            }

            // the ctags command must be executed if --ctags is specified with an empty string
            ctagsManager = boost::shared_ptr<CTagsManager>(new CTagsManager(
                    ctags_file, ctags, ctags != "", refposition));
            sourcehighlight.setCTagsManager(ctagsManager.get());
        }

        if (args_info.range_separator_given) {
            sourcehighlight.setRangeSeparator(args_info.range_separator_arg);
        }

        // OK, let's highlight!

        if (args_info.inputs_num && (args_info.input_given
                || args_info.output_given)) {
            // do not mix command line invocation modes
            cerr << "Please, use one of the two syntaxes for invocation: "
                    << endl;
            cerr
                    << "source-highlight [OPTIONS]... -i input_file -o output_file"
                    << endl;
            cerr << "source-highlight [OPTIONS]... [FILES]..." << endl;
            exit(EXIT_FAILURE);
        }

        // for cgi we can only process one file specified with --input
        if (args_info.inputs_num && !is_cgi) {
            // the output file name is empty, but we don't want to generate to stdout
            sourcehighlight.setCanUseStdOut(false);
            // in case multiple input files were specified (without --input)
            for (unsigned int i = 0; i < (args_info.inputs_num); ++i) {
                PROGRESSINFO(string("Processing ")+ args_info.inputs[i] + " ... ");
                sourcehighlight.highlight(args_info.inputs[i], "",
                        getLangFileName(inferPolicy, args_info.inputs[i],
                                langFile, srcLang, langmap));
                PROGRESSINFO("created " + sourcehighlight.createOutputFileName(args_info.inputs[i]) + "\n");
            }
        } else {
            if (is_cgi)
                print_cgi_header();

            // in case only one input file was specified with --input
            sourcehighlight.highlight(inputFile, outputFile, getLangFileName(
                    inferPolicy, inputFile, langFile, srcLang, langmap));
        }
    } catch (const HighlightBuilderException &e) {
        cerr << e << endl;
        exit(EXIT_FAILURE);
    } catch (const ParserException &e) {
        cerr << e << endl;
        exit(EXIT_FAILURE);
    } catch (const exception &e) {
        exitError(e.what());
    }

    cmdline_parser_free(&args_info);

    return 0;
}

void print_copyright() {
    int i;

    for (i = 1; i <= copyright_text_length; ++i)
        cout << copyright_text[i] << endl;;
}

void print_reportbugs() {
    int i;

    for (i = 1; i <= reportbugs_text_length; ++i)
        cout << reportbugs_text[i] << endl;
}

void print_version() {
    cout << Versions::getCompleteVersion() << endl;
}

string inferLang(const string &inputFileName) {
    VERBOSELN("inferring input language...");
    if (!inputFileName.size()) {
        printError("missing feature: language inference requires input file");
        return "";
    }

    LanguageInfer languageInfer;

    const string &result = languageInfer.infer(inputFileName);
    if (result.size()) {
        VERBOSELN("inferred input language: " + result);
    } else {
        VERBOSELN("couldn't infer input language");
    }

    return result;
}

string getMappedLang(const string &s, LangMap &langmap) {
    // OK now map it into a .lang file
    string mapped_lang = langmap.getMappedFileName(s);

    if (!mapped_lang.size()) {
        // try the lower version
        mapped_lang = langmap.getFileName(Utils::tolower(s));
    }

    return mapped_lang;
}

string getLangFileName(InferPolicy infer, const string &inputFileName,
        const string &langFileName, const string &langName, LangMap &langMap) {
    string langFile;

    // language inference has the precedence
    if (infer == INFERFIRST) {
        langFile = inferLang(inputFileName);
        langFile = getMappedLang(langFile, langMap);
        if (langFile.size())
            return langFile;
    }

    // then if a lang file name was specified we're done
    if (langFileName.size())
        return langFileName;

    // now try with the langName
    langFile = getMappedLang(langName, langMap);
    if (langFile.size())
        return langFile;

    // otherwise try with the inputFileName (its file extension
    // and its name)
    langFile = langMap.getMappedFileNameFromFileName(inputFileName);
    if (langFile.size())
        return langFile;

    // OK, as a last chance let's try with language infer
    if (infer == INFERATLAST) {
        langFile = getMappedLang(inferLang(inputFileName), langMap);
        if (langFile.size())
            return langFile;
    }

    // if we're here we failed all checks
    // if failsafe is specified we default to default.lang
    if (failsafe)
        langFile = "default.lang";

    if (!langFile.size()) {
        // if we're here we must exit with failure
        if (langName.size())
            exitError("could not find a language definition for " + langName);
        else
            exitError("could not find a language definition for input file "
                    + inputFileName);
    }

    return langFile;
}

void printError(const string &s) {
    cerr << PACKAGE << ": " << s << endl;
}

void exitError(const string &s) {
    printError(s);
    exit(EXIT_FAILURE);
}

void print_cgi_header() {
    cout << "Content-type: text/html\n";
    cout << "\n";
}

